dateutil/parser.py

```diff
@@ -725,11 +726,20 @@ def _parse(self, timestr, dayfirst=None, yearfirst=None, fuzzy=False,
                     # Token is a number
                     len_li = len(l[i])
 
-                    if (len(ymd) == 3 and len_li in (2, 4) and
-                        res.hour is None and
-                        (i + 1 >= len_l or
-                         (l[i + 1] != ':' and
-                          info.hms(l[i + 1]) is None))):
+                    hms_idx = _find_hms_idx(i, l, info, allow_jump=True)
+                    if hms_idx is not None:
```

Comment:
> Is it because you need to get and store the results of _find_hms_idx within the if call, so you've moved it to the front?

Yes.  I don't especially care about the ordering, but I do really like having all of these cases be `elif`s.  This is easy to move back to its original location.  I think the much simpler solution is to be willing to call `_find_hms_idx` twice.  It can be moved back down to after the `elif len_li in (8, 12, 14):` block and changed to:

```
if _find_hms_idx(i, l, info, allow_jump=True) is not None:
    hms_idx = _find_hms_idx(i, l, info, allow_jump=True)
    [...]
```

This is a pretty cheap function call.  Let's at least do some profiling before worrying about caching (much less mutability and thread-safety).